File: fastApi-agent-service/backend/middleware/verify_signature.py
import hmac
import hashlib
from fastapi import Request, HTTPException
from starlette.middleware.base import BaseHTTPMiddleware
import json
import os
import logging

from backend.config import main as config

logger = logging.getLogger(__name__)


HMAC_SECRETS = {
    "agent": config.HMAC_AGENT_KEY,
    "user": config.HMAC_USER_KEY,
}

async def verify_signature(request: Request):
    """
    Verify HMAC signature for secure inter-service communication
    Enhanced security for health data transmission between services
    
    Headers required:
    - wellness-signature: HMAC signature
    - wellness-origin: Source service (agent/user)  
    - wellness-validate: Validation type (body/query/both)
    - wellness-timestamp: Request timestamp for replay protection
    """
    headers = request.headers

    signature = headers.get("wellness-signature")
    origin = headers.get("wellness-origin") 
    validate = headers.get("wellness-validate")
    timestamp = headers.get("wellness-timestamp")
    
    
    logger.info(f"Inter-service signature verification attempt from origin: {origin}")

    if not all([signature, origin, validate, timestamp]):
        logger.error(f"Incomplete signature headers from {origin}: signature={bool(signature)}, validate={bool(validate)}, timestamp={bool(timestamp)}")
        raise HTTPException(status_code=403, detail="Signature Invalid - Missing Headers")

    
    body_bytes = await request.body()
    try:
        body_json = json.loads(body_bytes.decode("utf-8")) if body_bytes else {}
    except json.JSONDecodeError:
        logger.warning("Failed to decode request body JSON")
        body_json = {}

    query_params = dict(request.query_params)

    
    if validate == "query":
        validator = json.dumps(query_params, separators=(',', ':')) + timestamp
    elif validate == "both":
        validator = json.dumps(query_params, separators=(',', ':')) + json.dumps(body_json, separators=(',', ':')) + timestamp
    else: 
        validator = json.dumps(body_json, separators=(',', ':')) + timestamp

   
    key = HMAC_SECRETS.get(origin)
    if not key:
        logger.error(f"Unknown origin attempting inter-service communication: {origin}")
        raise HTTPException(status_code=403, detail="Signature Invalid - Unknown Origin")

    
    generated_signature = hmac.new(
        key.encode("utf-8"),
        validator.encode("utf-8"),
        hashlib.sha256
    ).hexdigest()

    
    if not hmac.compare_digest(generated_signature, signature):
        logger.error(f"Signature mismatch from {origin} - potential security breach")
        raise HTTPException(status_code=403, detail="Signature Invalid - Authentication Failed")

   
    try:
        request_timestamp = int(timestamp)
        current_timestamp = int(datetime.now().timestamp() * 1000)
        time_diff = abs(current_timestamp - request_timestamp)
        
        
        if time_diff > 300000:  
            logger.warning(f"Request timestamp outside acceptable window from {origin}: {time_diff}ms")
            raise HTTPException(status_code=403, detail="Signature Invalid - Timestamp Expired")
            
    except (ValueError, TypeError):
        logger.error(f"Invalid timestamp format from {origin}: {timestamp}")
        raise HTTPException(status_code=403, detail="Signature Invalid - Invalid Timestamp")

    logger.info(f"Signature verification successful for {origin}")
    return True

class HealthDataSecurityMiddleware(BaseHTTPMiddleware):
    """
    Middleware for enhanced health data security and audit logging
    """
    
    async def dispatch(self, request: Request, call_next):
        
        user_agent = request.headers.get("user-agent", "Unknown")
        client_ip = request.client.host if request.client else "Unknown"
        
        logger.info(f"Health data request: {request.method} {request.url.path} from {client_ip} ({user_agent})")
        
       
        sensitive_endpoints = [
            "/api/internal/create-health-plan",
            "/api/internal/update-health-plan", 
            "/api/user/health-plan",
            "/api/user/progress",
            "/api/user/health-data"
        ]
        
        is_sensitive = any(endpoint in str(request.url.path) for endpoint in sensitive_endpoints)
        
        if is_sensitive:
            logger.info(f"Sensitive health data endpoint accessed: {request.url.path}")
            
           
            response = await call_next(request)
            response.headers["X-Content-Type-Options"] = "nosniff"
            response.headers["X-Frame-Options"] = "DENY"
            response.headers["X-XSS-Protection"] = "1; mode=block"
            response.headers["Cache-Control"] = "no-store, no-cache, must-revalidate, private"
            response.headers["Pragma"] = "no-cache"
            
            return response
        
        return await call_next(request)

from datetime import datetime

async def verify_health_data_access(request: Request, user_id: str = None):
    """
    Additional verification layer for health data access
    Ensures proper authentication and logs access for compliance
    """
    
    auth_header = request.headers.get("authorization")
    if not auth_header or not auth_header.startswith("Bearer "):
        logger.error(f"Health data access attempted without proper authentication")
        raise HTTPException(status_code=401, detail="Authentication required for health data access")
    
    
    access_time = datetime.now().isoformat()
    logger.info(f"Health data accessed by user {user_id} at {access_time}")
    
    
    request_count_key = f"health_access_{user_id}_{datetime.now().strftime('%Y-%m-%d-%H')}"
    
    
    
    return True

def validate_health_plan_request(request_data: dict) -> dict:
    """
    Validate health plan request data for safety and completeness
    """
    validation_result = {
        "is_valid": True,
        "errors": [],
        "warnings": []
    }
    
    
    required_fields = ["user_id", "primary_goal", "current_activity_level"]
    for field in required_fields:
        if field not in request_data or not request_data[field]:
            validation_result["is_valid"] = False
            validation_result["errors"].append(f"Required field missing: {field}")
    
    
    age = request_data.get("age")
    if age:
        if age < 13:
            validation_result["is_valid"] = False
            validation_result["errors"].append("Age below minimum safety threshold")
        elif age < 18:
            validation_result["warnings"].append("Minor - parental supervision recommended")
        elif age > 75:
            validation_result["warnings"].append("Advanced age - medical consultation strongly recommended")
    
    
    health_conditions = request_data.get("health_conditions", [])
    high_risk_conditions = [
        "heart", "cardiac", "diabetes", "blood pressure", "eating disorder",
        "pregnancy", "surgery", "injury", "medication"
    ]
    
    for condition in health_conditions:
        if any(risk in condition.lower() for risk in high_risk_conditions):
            validation_result["warnings"].append(f"Health condition requires medical consultation: {condition}")
    
    
    medical_clearance = request_data.get("medical_clearance", False)
    if validation_result["warnings"] and not medical_clearance:
        validation_result["warnings"].append("Medical clearance recommended before proceeding")
    
    return validation_result

File: nodeJs-user-service/middlewares/verifySignature.js
import crypto from 'crypto'
import config from '@/config'
import isEmpty from 'is-empty'
import { logAuthEvent } from '@/security/security'


export const verifySignature = async (req, res, next) => {
    try {
        const { headers, body, query } = req
        
        const signature = headers['wellness-signature']
        const origin = headers['wellness-origin']
        const validate = headers['wellness-validate']
        const timestamp = headers['wellness-timestamp']
        const isHealthData = headers['wellness-health-data'] === 'true'

        if (isEmpty(signature) || isEmpty(origin) || isEmpty(validate) || isEmpty(timestamp)) {
            logAuthEvent('hmac_missing_headers', null, false, {
                ipAddress: req.ip,
                userAgent: req.get('User-Agent'),
                additionalData: { 
                    missingHeaders: {
                        signature: isEmpty(signature),
                        origin: isEmpty(origin),
                        validate: isEmpty(validate),
                        timestamp: isEmpty(timestamp)
                    }
                }
            })

            return res.status(403).json({ 
                success: false, 
                message: 'Invalid signature headers - service authentication failed',
                error_type: 'signature_invalid',
                required_headers: ['wellness-signature', 'wellness-origin', 'wellness-validate', 'wellness-timestamp']
            })
        }

     
        const currentTime = Date.now()
        const requestTime = parseInt(timestamp)
        const timeDifference = Math.abs(currentTime - requestTime)
        const maxAllowedDifference = 5 * 60 * 1000 

        if (timeDifference > maxAllowedDifference) {
            logAuthEvent('hmac_timestamp_invalid', null, false, {
                ipAddress: req.ip,
                userAgent: req.get('User-Agent'),
                additionalData: { 
                    timeDifference: timeDifference / 1000,
                    requestTime: new Date(requestTime).toISOString(),
                    currentTime: new Date(currentTime).toISOString()
                }
            })

            return res.status(403).json({ 
                success: false, 
                message: 'Request timestamp invalid - possible replay attack',
                error_type: 'timestamp_invalid'
            })
        }

        let validator = JSON.stringify(body) + timestamp

        if (validate === 'query') {
            validator = JSON.stringify(query) + timestamp
        } else if (validate === 'both') {
            validator = JSON.stringify(query) + JSON.stringify(body) + timestamp
        }

     
        let key = ''
        if (origin === 'agent') {
            key = config.HMAC_AGENT_KEY
        } else if (origin === 'user') {
            key = config.HMAC_USER_KEY
        } else {
            logAuthEvent('hmac_invalid_origin', null, false, {
                ipAddress: req.ip,
                userAgent: req.get('User-Agent'),
                additionalData: { invalidOrigin: origin }
            })

            return res.status(403).json({ 
                success: false, 
                message: 'Invalid service origin',
                error_type: 'invalid_origin',
                allowed_origins: ['agent', 'user']
            })
        }

        if (!key) {
            console.error('[WELLNESS-HMAC] Missing HMAC key for origin:', origin)
            return res.status(500).json({ 
                success: false, 
                message: 'Service configuration error',
                error_type: 'configuration_error'
            })
        }

        const generatedSignature = crypto.createHmac('SHA256', Buffer.from(key, 'utf-8'))
        generatedSignature.update(validator)
        const latestSignature = generatedSignature.digest('hex')

        let isValid = false
        try {
            isValid = crypto.timingSafeEqual(
                Buffer.from(latestSignature, 'hex'), 
                Buffer.from(signature, 'hex')
            )
        } catch (error) {
            logAuthEvent('hmac_signature_format_error', null, false, {
                ipAddress: req.ip,
                userAgent: req.get('User-Agent'),
                additionalData: { error: error.message }
            })

            return res.status(403).json({ 
                success: false, 
                message: 'Signature format invalid',
                error_type: 'signature_format_error'
            })
        }

        if (!isValid) {
            logAuthEvent('hmac_signature_mismatch', null, false, {
                ipAddress: req.ip,
                userAgent: req.get('User-Agent'),
                additionalData: { 
                    origin,
                    validate,
                    isHealthData,
                    bodySize: JSON.stringify(body).length,
                    queryParams: Object.keys(query).length
                }
            })

            return res.status(403).json({ 
                success: false, 
                message: 'Signature verification failed - service authentication invalid',
                error_type: 'signature_mismatch'
            })
        }

     
        if (isHealthData) {
            console.log(`[WELLNESS-HMAC] Health data request authenticated from ${origin} service`)
        }

        req.signatureVerified = {
            origin,
            timestamp: requestTime,
            isHealthData,
            verifiedAt: new Date()
        }

       
        logAuthEvent('hmac_verification_success', null, true, {
            ipAddress: req.ip,
            userAgent: req.get('User-Agent'),
            additionalData: { 
                origin,
                validate,
                isHealthData,
                endpoint: req.path
            }
        })

        return next()

    } catch (error) {
        console.error('[WELLNESS-HMAC] Signature verification error:', error)
        
        logAuthEvent('hmac_verification_error', null, false, {
            ipAddress: req.ip,
            userAgent: req.get('User-Agent'),
            additionalData: { error: error.message }
        })
        
        return res.status(500).json({ 
            success: false, 
            message: 'Signature verification service error. For immediate health concerns, please consult healthcare professionals.',
            error_type: 'signature_service_error'
        })
    }
}


export const validateHealthPlanRequest = (requestData) => {
    const validation = {
        is_valid: true,
        errors: [],
        warnings: []
    }


    if (requestData.age && (requestData.age < 13 || requestData.age > 100)) {
        validation.is_valid = false
        validation.errors.push('Age must be between 13 and 100 years')
    }

    const highRiskConditions = [
        'heart disease', 'diabetes', 'high blood pressure', 'eating disorder',
        'pregnancy', 'recent surgery', 'joint problems', 'back injury',
        'chest pain', 'shortness of breath'
    ]

    if (requestData.health_conditions) {
        const hasHighRiskCondition = requestData.health_conditions.some(condition =>
            highRiskConditions.some(risk => 
                condition.toLowerCase().includes(risk.toLowerCase())
            )
        )

        if (hasHighRiskCondition) {
            validation.warnings.push('High-risk health conditions detected - professional consultation strongly recommended')
        }
    }

    if (requestData.primary_goal === 'extreme_weight_loss' || 
        requestData.time_availability_minutes > 180) {
        validation.warnings.push('Extreme fitness goals detected - medical supervision recommended')
    }

    if (requestData.age > 65 || requestData.health_conditions?.length > 0) {
        if (!requestData.medical_clearance) {
            validation.warnings.push('Medical clearance recommended before starting program')
        }
    }

    return validation
}


export const healthEndpointRateLimit = async (req, res, next) => {
    const clientIP = req.ip
    const userId = req.user?._id?.toString()
    const endpoint = req.path
    
  
    const rateLimitKey = `health_rate_limit:${clientIP}:${userId}:${endpoint}`

    console.log(`[WELLNESS-RATE-LIMIT] Health endpoint access: ${endpoint} from ${clientIP}`)
  
    next()
}